### PR TITLE
PrePi: Fix FDT parsing

### PR DESCRIPTION
**Pull request by feherneoh**
All files: Add inheritance to #address-cells and #size-cells
PrePi/FdtParser.c: Add support for /reserved-memory FDT node

### CODE REVIEW

File: PrePi/FdtParser.c [hunk 1]
```diff
@@ -91,6 +118,67 @@ FindMemnodeFdt (
...
+      Prop += AddressCells;
+
+      LocalSize = fdt32_to_cpu (Prop[0]);
+      if (SizeCells > 1) {
+        LocalSize = (LocalSize << 32) | fdt32_to_cpu (Prop[1]);
+      }
+      Prop += SizeCells;
+
+      if (HighestNoMap < LocalBase + LocalSize) {
+        HighestNoMap = LocalBase +LocalSize;
```
Comment (M1cha) [inline PrePi/FdtParser.c]: space after +

File: PrePi/FdtParser.c [hunk 2]
```diff
@@ -22,17 +22,30 @@ FindMemnodeFdt (
   )
 {
   INT32         MemoryNode;
+  INT32         RootNode;
+  INT32         ReservedMemoryNode;
   INT32         AddressCells;
   INT32         SizeCells;
+  INT32         AddressCellsRoot;
+  INT32         SizeCellsRoot;
```
Comment (M1cha) [inline PrePi/FdtParser.c:30]: why are you storing them separately instead of just overriding the existing ones?

Comment (feherneoh) [reply PrePi/FdtParser.c:30]: The root ones are reused later, that's why they are stored separately

Comment (M1cha) [reply PrePi/FdtParser.c:30]: I know but you're copying them to the normal ones anyway.

File: PrePi/FdtParser.c [hunk 3]
```diff
@@ -43,10 +56,24 @@ FindMemnodeFdt (
...
   //
   AddressCells = 1;
   SizeCells = 1;
+  AddressCellsRoot = -1;
+  SizeCellsRoot = -1;
+
+  Prop = fdt_getprop (DeviceTreeBlob, RootNode, "#address-cells", &Length);
+  if (Length == 4) {
+    AddressCells = fdt32_to_cpu (*Prop);
+    AddressCellsRoot = AddressCells;
```
Comment (M1cha) [inline PrePi/FdtParser.c:69]: at this point AddressCells == AddressCellsRoot, I don't get the reason.

File: PrePi/FdtParser.c [hunk 4]
```diff
@@ -91,6 +118,67 @@ FindMemnodeFdt (
     }
   }
 
+  //
+  // no-map and no-map-fixup
+  //
+  HighestNoMap = 0;
+  AddressCells = (AddressCellsRoot == -1) ? 1 : AddressCellsRoot;
```
Comment (M1cha) [inline PrePi/FdtParser.c:125]: if the root node wasn't found you'll override AddressCells to 1 even though it's either still 1 or got set by the local node.
If the root node was found you use AddressCellsRoot no matter if the local one has been found or not.

Comment (feherneoh) [reply PrePi/FdtParser.c:125]: Line 134-137 overrides it if a local value is found in the other node checked

File: PrePi/FdtParser.c (hunk 3 above)
Comment (feherneoh) [reply PrePi/FdtParser.c:69]: The reason is that this way it does not have to be copied later if the local value is not found

File: PrePi/FdtParser.c (hunk 2 above)
Comment (feherneoh) [reply PrePi/FdtParser.c:30]: It gets copied to the normal ones, but if it wasn't stored separately, one would have to read the value again before using it as a default for the second node checked.
There is only one root node, but the code checks 2 childnodes.

**Pull request merged**